---
title: "Kryazhimskiy data analysis notebook"
author: Kyle Card
Date: 3/27/2020
output: html_notebook
---

## Prerequisites

Clear current environment
```{r Clear current environment, messages = FALSE}
rm(list = ls())
```
<br>

Load packages for use
```{r Packages, messages = FALSE}
library(tidyverse)
library(reshape2)
library(proxy)
```
<br>

***

```{r Reads in data}
# Binary data showing whether a mutation is present (1) or absent (0) at a given locus across all replicate lines (columns) 
specificity_bg <- read.csv("Kryazhimskiy_mutation_matrix.csv", stringsAsFactors = FALSE)
```
<br>

***

# Specificity of genomic evolution with respect to background in the Kryazhimskiy data

We quantify the extent of parallelism in genome evolution between lines evolved from the same parental genotype and across different parental genotypes. For these calculations, we included only those mutations that were analyzed in their study.

We estimate Dice's coefficient of similarity, *S*, for each pair of replicate lines, where $$S = \frac{2|X \cap Y|}{|X|+|Y|}$$ $|X|$ and $|Y|$ are the sets of genes with mutations in two evolved clones, and $|X \cap Y|$ is the set of genes with mutations in both clones. *S* therefore ranges from 0, when the pair of backgrounds share no mutations in common, to 1, when both have mutations in exactly the same set of genes (Deatherage et al., 2017).
```{r Specificity function, messages = FALSE}
Specificity <- function(dat, randomization = FALSE) {
    mutations <- dat %>% 
      select(!gene) # Removes the "gene" column

  if (randomization == TRUE) {
    colnames(mutations) <- sample(colnames(mutations)) # Randomly shuffles label names but keeps the identity of mutations in a given column the same
    }
  
  # Computes similarity coefficient for all possible replicate pairs. Populates a matrix with these values
  pwise_simil_mat <- as.matrix(simil(mutations, method = "Dice", by_rows = FALSE, upper = TRUE, diag = TRUE))
  
  # All values above (and including) the matrix diagonal are converted to NA to ease downstream wrangling
  pwise_simil_mat[upper.tri(pwise_simil_mat)] <- NA
  
  # Converted matrix to a data frame for analysis. NA values above (and including) the matrix diagonal are incorporated into the newly formed dataframe. This piece of code drops the rows contanining NA, effectively retaining only those values *below* the matrix diagonal
  pwise_simil_df <- melt(pwise_simil_mat, varnames = c("background_1", "background_2"))
  
  df_filtered <- pwise_simil_df %>% 
    drop_na()
  
  # This code removes the prefixes or suffixes from the clone labels to ease downstream wrangling
  background_vec <- c("background_1", "background_2")
  
  df_filtered <- df_filtered %>%
    mutate_at(
      .vars = vars(background_vec),
      .funs = ~ str_replace_all(., pattern = c("_1|_2|_3|_4|_5|_6|_7|_8|_9|_10|_11|_12|_13"), replacement = "")
      )
  
  
  # First part of the summary output of the Specificity function if not performing subsequent randomization test
  avg_simils_df <- df_filtered %>% 
    group_by(background_1, background_2) %>% 
    summarize(avg = mean(value)) %>% 
    as_tibble()
  
  # This next set of code pools clones into two groups: i.) those that evolved from the same parental genotype (S_s); and ii.) those that evolved from different parental genotypes (S_d)
  genotype_vec <- c("L013", "L048", "L098", "L102", "L041", "L094", "L096a", "L003", "S121", "S002", "S028", "L096b", "L102a")
  simil_same_df <- map_df(genotype_vec, ~ filter(df_filtered, background_1 == .x & background_2 == .x))
  
  simil_diff_df <- anti_join(df_filtered, simil_same_df)
  
  weighted_avg_same <- simil_same_df %>% 
    summarize(weight_avg = mean(value)) %>% 
    round(digits = 5)
  
  weighted_avg_diff <- simil_diff_df %>% 
    summarize(weight_avg = mean(value)) %>% 
    round(digits = 5)
  
  bound_weighted_avgs <- bind_rows(weighted_avg_same, weighted_avg_diff)
  
  
  # Second part of the summary output of the Specificity function if not performing a subsequent randomization test
  gen_group_col <- data.frame(genotype = c("Same", "Different"))
  weighted_avgs_df <- bind_cols(gen_group_col, bound_weighted_avgs)
  

  # Third part of the summary output of the Specificity function if not performing a subsequent randomization test
  specificity <- weighted_avg_same - weighted_avg_diff
  names(specificity) <- "specificity"
  
  # If the Specificity function is told to perform a randomization test, then only the specificity statistic is returned. The significance of this statistic is then calculated. 
  if (randomization == FALSE) {
    results <- list(avg_simils_df, weighted_avgs_df, specificity)
  } else {
    results <- specificity
  }

  return(results)
}
```
<br>

By default, the Specificity function will output a list containing three elements: i.) average similarity among all clone pairs, either evolved from the same parental genotype or between two parental genotypes; ii.) pooled averages; and iii.) the difference between these pooled averages, which is a measure of specificity of genomic evolution with respect to genetic background. Specifically, this calculation yields a positive or negative value that indicates the distinctness of clones evolved from the same parental genotype relative to backgrounds from different parental genotypes. Similarity is higher in the former group and lower in the latter with increasing positive values. The converse is true if the value is negative (Sokal and Rohlf, 1994).
```{r Summary by genetic background}
background_summary <- Specificity(specificity_bg)
print(background_summary)
```
<br>

If we consider an observed specificity statistic as one of many possible but equally likely different outcomes that could have arisen by chance, we can perform a randomization test to evaluate its significance. To do this, we randomly rearrange the clone labels, but retain the number and identity of mutations in any clone. A similarity statistic is calculated for each trial and these outcomes are enumerated. Lastly, we quantify the proportion of trials that have a similarity statistic equal to, or greater than that observed (i.e., approximate *p*-value).
```{r Randomization test function, messages = FALSE}
# By default the RandomizationTest function will perform 10,000 trials. This value may be changed based upon your needs.
RandomizationTest <- function(dat, trials = 10000) {
  trial_specificity <- c()
  
  observed_specificity <- Specificity(dat)
  observed_specificity <- observed_specificity[[3]]$specificity
  
  for (i in 1:trials) {
    trial_specificity[i] <- Specificity(dat, randomization = TRUE)
  }
  
  significance <- sum(trial_specificity >= observed_specificity) / trials # This equation gives the approximate p-value
    
  return(significance)
}
```
<br>

```{r Significance of specificity with respect to genetic background, messages = FALSE}
background_sig <- suppressMessages(RandomizationTest(specificity_bg))
print(background_sig)
```
<br>
















