---
title: "LTEE ABR mutant sequencing notebook"
author: Kyle Card
Date: 3/1/2020
output: html_notebook
---

## Prerequisites

Clear current environment
```{r Clear current environment, messages = FALSE}
rm(list = ls())
```
<br>

Load packages for use
```{r Packages, messages = FALSE}
library(tidyverse)
library(reshape2)
library(proxy)
library(metap) # Might need to first (i.) install.packages("BiocManager), and then (ii.) run BiocManager::install("multtest") 
library(colorspace)
library(cowplot)
```
<br>

***

```{r Factor levels}
antibiotics <- c("AMP", "CRO", "CIP", "TET")
antibiotic_labels <- c(
  `AMP` = "Ampicillin",
  `CRO` = "Ceftriaxone",
  `CIP` = "Ciprofloxacin",
  `TET` = "Tetracycline"
)
mutation_type <- c("Nonsynonymous","Nonsense","Synonymous", "Intergenic", "Indel < 50 bp", "Deletion > 50 bp", "IS insertion", "Other")
functional_levels <- c("Gene regulation", "Biosynthesis or metabolism", "Transport", "Transcription or translation", "Replication")
```
<br>

***

# Data import and wrangling

```{r Read in data, messages = FALSE}
# Reads in the reference sequence data of LTEE clones (Ara-5, Ara-6, Ara+4, Ara+5) at generation 50,000 (obtained from Tenaillon et al., 2016).
reference_sequences_df <- read.csv("References.csv", stringsAsFactors = FALSE)

# Reads in the mutant clone sequence data (obtained from this study).
sample_sequences_df <- read.csv("breseq output.csv", stringsAsFactors = FALSE)
```
<br>

***

# Curation of *breseq* results

Reads corresponding to resistant mutants evolved from the LTEE ancestral and generation 50,000 derived clones (i.e., Ara-5, Ara-6, Ara+4, and Ara+5) were mapped to the REL606 reference genome using *breseq v. 0.35.0*. 

Each sequenced resistant isolate contains information on background and resistance mutations that evolved during the LTEE and our previous selection experiments (Card et al., 2019), respectively. We discarded the background mutations by comparing each evolved isolate to its corresponding parental strain (Tenaillon et al., 2016). For example, isolate **KJC64** is a resistant mutant evolved from the **Ara+4 background at 50,000**. We therefore compared its genome against its parental strain. If both strains have a mutation at the same genome position, then it is assumed that this is a background mutation and is discarded from further analyses. We further curated this list of mutational differences (**mutations_df** object) by removing predicted sample mutations in genome repeats and insertion sequence regions (e.g., *hemX* or *insF*, respectively). We also discarded point mutations that lie in deleted regions of the reference genome and mutations in multi-copy number genes (i.e., *rhsB*, *rrfC*, *rrlB*). Lastly, we found that mutant clones **KJC152, KJC184, and KJC217** were cross-contaminants using the *gdtools_analysis* script. We discarded them from all further analyses.

Using the *gdtools* utility, we also note several examples where the parental strain's background contains a mutation that is absent in some, or all, evolved resistant daughter backgrounds: an amplification *[ascF]*-*insJ-3* is found in the **Ara+5 parental background** but it collapsed (isn't present) in all evolved daughter backgrounds; the *[yjcC]*-*[alsB]* / *[yjcF]*-*[alsB]* amplications found in the **Ara+5 parental background** collapsed in the **KJC65**, **KJC121**, **KJC153**, **KJC185** daughter mutants; *[fhlA]*-*insJ‑3* duplication is found in the **Ara-5 parental background** but it collapsed in all evolved daughter backgrounds
```{r Filtered mutations}
CompareMutations <- function(ref_dat, mutant_dat) {
  bound_df <- c()
  
  for (pop in unique(ref_dat$population)) {
    ref_subset <- ref_dat %>% 
      filter(population == pop)
    
    mutant_pop_subset <- mutant_dat %>% 
      filter(population == pop)
    
    for (strain in unique(mutant_pop_subset$designation)) {
      mutant_subset <- mutant_pop_subset %>% 
        filter(designation == strain)
      
      bound_df <- bind_rows(bound_df, anti_join(mutant_subset, ref_subset, by = "position"))
    }
  }
  return(bound_df)
}

mutations_df <- CompareMutations(reference_sequences_df, sample_sequences_df)

mutations_df <- mutations_df %>% 
  filter(!gene %in% c("ybcW\U2013[ECB_01526]","[rbsD]","rbsD\u2013[rbsK]","araA", "cpsG", "hemX", "insF-1", "manB", "recD", "rhsB", "rrfC")) %>% 
  filter(!designation %in% c("KJC152", "KJC184", "KJC217")) %>% 
  arrange(id)

mutations_df$id <- 1:nrow(mutations_df)

# write.csv(mutations_df, file = "mutations.csv", na = "NA", col.names = TRUE, row.names = FALSE) 
```
<br>

*breseq* did not predict resistance mutations in 11 strains (**KJC64**, **KJC65**, **KJC66**, **KJC67**, **KJC83**, **KJC115**, **KJC120**, **KJC123**, **KJC131**, **KJC161**, **KJC185**). These strains may have unresolved amplifications that conferred phenotypic resistance (e.g., increased copy number of efflux pumps or target proteins), and/or "unassigned" missing coverage or new junction evidence that would not be caught in the above filtering protocol. Therefore, we manually examined the *breseq* outputs for this evidence, including outputs for signatures of genome duplications by visualizing the read coverage for these strains and looking for genome regions that have above-average coverage where no called mutations exist. We also verified mutation predictions by using the *gdtools* utility to apply mutations from the parental backgrounds (determined in Tenaillon et al. 2016) to the REL606 reference genome and reran *breseq*.  

Using this approach, we resolved mutations in 9 of the above strains. The details of these mutations are as follows: **KJC64** has an IS insertion in *rhlB*; **KJC67** has a large (~260,278 bp) amplification from *[yfcTu]* to *[yfiH]*; **KJC83** and **KJC115** have a large (~456,332 bp) amplification from *yeaS* to *menC*; **KJC120** has a large (~1,052,344 bp) amplification from *lysU* to *aroB*; **KJC123** and **KJC131** have a large (1,446 bp) deletion in *yfiH*; and **KJC161** and **KJC185** have large deletions (~24,000 bp) between *yedV* and *cbl/nac*. All of these mutations are mediated by insertion sequence elements. Unfortunately, we were not able to resolve mutations in **KJC65** and **KJC66**. These strains will be described as having "Unknown" mutation(s).

Moreover, we also discovered *new* mutations in 9 other strains. The details of these mutations are as follows: **KJC61** has a large (~204,077 bp) amplification from *[insB-4]* to *ECB_00515*; **KJC74** has a large amplification (~635,224 bp) amplification between *insB-1* and *gltL*; **KJC75** has a large (~260,278 bp) amplification from *yfcTu* to *yfiH*; **KJC81** has an IS insertion in *yfcV*; **KJC114** has a large (~740,483 bp) amplification from *insB-28* to *gltL*; **KJC218** and **KJC234** have a large (~16,049 bp) amplification from *insA-26* to *fimB*; and **KJC219** and **KJC235** have a large (~456,332 bp) amplification from *yeaS* to *menC*.

There are three interesting observations from these data. First, the same amplification of *yeaS* to *menC* occurs in 4/12 strains that evolved from the Ara-6 background; these 4 strains evolved in three different antibiotics. Second, there is parallelism among two mutants evolved from the **Ara-5** background in ceftriaxone: they both have an amplification from *insA-26* to *fimB*. Third, there is parallelism among two mutants evolved from the **Ara+5** background to the drug ciprofloxacin: they both have deletions of *yedV* to *cbl/nac*.

We manually changed the gene entry for **KJC150** from *dnaG -> / -> rpoD* to reflect the fact that this intergenic mutation occurs downstream of *dnaG* gene and within the promoter region of *rpoD*. Likewise, for **KJC216** we changed the gene entry from *hns <- / -> tdk* to *hns*. The mutation occurs 24 bp upstream of *hns* and 572 bp upstream of *tdk* and is therefore in the *hns* promoter region. For **KJC224** we changed *rrmA <- / <- cspC* to *rrmA* because the intergenic IS insertion is downstream of *cspC* and in the promoter of *rrmA*. For **KJC226** we changed *aspC <- / <- ompF* to *aspC* for the same reasons. 

We added all of this information manually to the outputted **mutations.csv** file from the previous code-chunk. We also added a binary "qualifying" mutation score (described later), mutation type, and biological process affected by the given mutation. We renamed this file **mutations_curated.csv**

**NOTE:** The *breseq* outputs and read coverage images for each strain can be found on Card's Github repository.
```{r Reads in curated data}
mutations_curated_df <- read.csv("Mutations_curated.csv", stringsAsFactors = FALSE)
```
<br>

***

# Binomial tests of the distributions of large deletions and amplifications

Large deletions and amplifications were predominantly found in lines evolved from derived backgrounds relative to lines evolved from the ancestral background. We examine the probability of these outcomes assuming the null hypothesis that these genetic changes are more predominant in derived backgrounds. For example, there are 16 genomes evolved from the LTEE ancestor and 45 evolved from the generation 50,000 clones. So, 45/61 = 73.77% of these mutations should occur in the derived backgrounds.
```{r Binomial tests}
binom_test <- function(dat, type) {
  n <- dat %>% 
    filter(mutation_type == type) %>% 
    nrow()
  
  k <- dat %>% 
    filter(mutation_type == type & population != "Ancestor") %>% 
    nrow()

  results <- binom.test(k, n, p = 0.7377, alternative = "greater")
  
  return(results)
}

del_amp_vec <- c("Deletion > 50 bp", "Amplification > 50 bp")
binom_test_results <- map(del_amp_vec, ~ binom_test(mutations_curated_df, type = .x))
names(binom_test_results) <- del_amp_vec

binom_test_results
```
<br>

***

# Genomic parallelism on the functional level

We quantify the extent of parallelism in genome evolution among all resistant lines by the cellular function affected. These functions include gene regulation, biosynthesis or metabolism, transport, transcription or translation, replication, and other. However, each functional unit contains a different number of genes, and therefore there is variation in mutational target sizes. To account for this fact, we model the data using a Poisson distribution.
```{r Poisson analysis, message = FALSE}
gene_lengths_df <- read_csv("gene_lengths.csv")

PoissonAnalysis <- function(gene_lengths, curated_mutations, proc, background = NULL) {
  genome_size <- (4629812 + 4593005 + 4541082 + 4670194 + 4578768) / 5 # Average genome size of parental strains to get an unbiased estimate of lambda
  
  target_size <- gene_lengths %>% # Mutational target size in bp
    filter(process == proc) %>% 
    summarize(total = sum(gene_length))
  
  # Removes large deletions or amplifications when the affected genes do not share a common role; also removes "Unknown" mutations. The remaining mutations (n = 57) are used in the analysis
  total_mutations <- curated_mutations %>%  
    filter(!process %in% c("Unknown", "Multiple")) %>% 
    nrow()
  
  observed <- curated_mutations %>% # Number of mutations that affect genes with a given function
      filter(process == proc) %>% 
      nrow()
  
  expected <- total_mutations * (target_size$total / genome_size) # Expected number of mutations given the mutational target size (i.e., lambda) (Tenaillon et al. 2012)

  
  prob <- ppois(q = (observed - 1), lambda = expected, lower.tail = FALSE) # P(x >= observed given lambda); a.k.a., p-value
  
  results <- data.frame(process = proc,
                            observed_mutations = observed,
                            expected_mutations = expected,
                            p_value = prob)
  
  return(results)
}

poisson_analysis <- map_df(functional_levels, ~ PoissonAnalysis(gene_lengths_df, mutations_curated_df, .x))


# Calculate expected mutations in the "Other" catagory
expected_other <- 57 - sum(poisson_analysis$expected_mutations)

# Calculate Poisson probability for "Other" catagory
prob_other <- ppois(q = (2 - 1), lambda = expected_other, lower.tail = FALSE)

# Generate data frame row with this information
poisson_analysis_other <- data.frame(process = "Other",
                                     observed_mutations = 2,
                                     expected_mutations = expected_other,
                                     p_value = prob_other)

# Combine data frames
poisson_analysis_final <- bind_rows(poisson_analysis, poisson_analysis_other)
poisson_analysis_final
```
<br>

***

# Specificity of genomic evolution with respect to genetic background and antibiotic treatment

We quantify the extent of parallelism in genome evolution among resistant mutants evolved from the same parental genotype and between different parental genotypes, or among resistant mutants within an antibiotic treatment and across different treatments. For these calculations, we included only those "qualifying" mutations that unambiguously impact a single gene, including non-synonymous point mutations, small indels (including IS elements), and mutations occuring in intergenic regions within 150 bp of the start of the gene. We excluded synonymous mutations, deletions and amplifications that overlap multiple genes (unless one of the genes was found to be mutated in a separate clone or if there was parallel changes across different lines) (Deatherage et al., 2017). We also omitted from our analysis the two lines with no clear genetic changes. A total of *71* mutations qualify based upon these criteria.

We estimate Dice's coefficient of similarity, *S*, for each pair of evolved backgrounds, where $$S = \frac{2|X \cap Y|}{|X|+|Y|}$$ $|X|$ and $|Y|$ are the sets of genes with mutations in two evolved clones, and $|X \cap Y|$ is the set of genes with mutations in both clones. *S* therefore ranges from 0, when the pair of backgrounds share no mutations in common, to 1, when both have mutations in exactly the same set of genes (Deatherage et al., 2017).
```{r Specificity function, messages = FALSE}
# Binary data showing whether a mutation is present (1) or absent (0) in a given locus across all replicates and treatments
specificity_bg <- read.csv("Specificity by background.csv", stringsAsFactors = FALSE)
specificity_ab <- read.csv("Specificity by antibiotic.csv", stringsAsFactors = FALSE)

# Specificity function
Specificity <- function(dat, ab = NULL, randomization = FALSE) {
  if (is.null(ab)) {
    mutations <- dat %>% 
      select(!c(gene, AraP4_1_CIP, AraP4_3_CIP, AraP5_1_CRO)) # Removes cross contaminants
  } else {
    mutations <- dat %>% 
      filter(antibiotic == ab) %>% 
      select(!(antibiotic:gene)) # Removes the first two columns, which are not needed in the function that estimates pairwise similarities
    
    if (ab == "CRO") { # Removes cross contaminant
      mutations <- mutations %>% 
        select(!AraP5_1)
    }
    
    if (ab == "CIP") { # Removes cross contaminants
      mutations <- mutations %>% 
        select(!c(AraP4_1, AraP4_3))
      }
    }

  if (randomization == TRUE) {
    colnames(mutations) <- sample(colnames(mutations))
    }
  
  # Computes similarity coefficient for all possible replicate pairs. Populates a matrix with these values
  pwise_simil_mat <- as.matrix(simil(mutations, method = "Dice", by_rows = FALSE, upper = TRUE, diag = TRUE))
  
  # All values above (and including) the matrix diagonal are converted to NA to ease downstream wrangling
  pwise_simil_mat[upper.tri(pwise_simil_mat)] <- NA
  
  # Converted matrix to a data frame for analysis. NA values above (and including) the matrix diagonal are incorporated into the newly formed dataframe. This piece of code drops the rows contanining NA, effectively retaining only those values *below* the matrix diagonal
  pwise_simil_df <- melt(pwise_simil_mat, varnames = c("background_1", "background_2"))
  
  df_filtered <- pwise_simil_df %>% 
    drop_na()
  
  # This code removes the prefixes or suffixes from the clone labels to ease downstream wrangling
  background_vec <- c("background_1", "background_2")
  
  if (is.null(ab)) {
    df_filtered <- df_filtered %>% 
      mutate_at(
        .vars = vars(background_vec),
        .funs = ~ str_replace_all(., pattern = c("_1_|_2_|_3_|_4_"), replacement = "_")
      ) %>% 
      mutate_at(
        .vars = vars(background_vec),
        .funs = ~ str_replace_all(., pattern = c("Anc_|AraM5_|AraM6_|AraP4_|AraP5_"), replacement = "")
      )
  } else {
    df_filtered <- df_filtered %>%
      mutate_at(
        .vars = vars(background_vec),
        .funs = ~ str_replace_all(., pattern = c("_1|_2|_3|_4"), replacement = "")
      )
  }
  
  
  # First part of the summary output of the Specificity function if not performing subsequent randomization tests
  avg_simils_df <- df_filtered %>% 
    group_by(background_1, background_2) %>% 
    summarize(avg = mean(value)) %>% 
    as_tibble()
  
  # This next set of code pools clones into two groups depending on the "ab" argument passed to the Specificity function. If one does not specify an antibiotic, then it is assumed that one wishes to evaluate the specificity of genomic evolution with respect to antibiotic treatment. The two groups in this case are: i.) clones from the same antibiotic (S_st); and ii.) clones from different antibiotics (S_dt). However, if one specifies an antibiotic (or multiple antibiotics), then it is assumed that one wishes to evaluate the specificity based upon genetic background. The two groups in this case are: i.) those that evolved fromt the same parental genotype (S_sg); and ii.) those that evolved from different parental genotypes (S_dg)
  if (is.null(ab)) {
    antibiotic_vec <- c("AMP", "CRO", "CIP", "TET")
    simil_same_df <- map_df(antibiotic_vec, ~ filter(df_filtered, background_1 == .x & background_2 == .x))
  } else {
    genotype_vec <- c("Anc", "AraM5", "AraM6", "AraP4", "AraP5")
    simil_same_df <- map_df(genotype_vec, ~ filter(df_filtered, background_1 == .x & background_2 == .x))
  }
  
  simil_diff_df <- anti_join(df_filtered, simil_same_df)
  
  weighted_avg_same <- simil_same_df %>% 
    summarize(weight_avg = mean(value)) %>% 
    round(digits = 5)
  
  weighted_avg_diff <- simil_diff_df %>% 
    summarize(weight_avg = mean(value)) %>% 
    round(digits = 5)
  
  bound_weighted_avgs <- bind_rows(weighted_avg_same, weighted_avg_diff)
  
  
  # Second part of the summary output of the Specificity function if not performing a subsequent randomization test
  if (is.null(ab)) {
    ab_group_col <- data.frame(antibiotic = c("Within", "Between"))
    weighted_avgs_df <- bind_cols(ab_group_col, bound_weighted_avgs)
  } else {
    gen_group_col <- data.frame(genotype = c("Same", "Different"))
    weighted_avgs_df <- bind_cols(gen_group_col, bound_weighted_avgs)
  }
  

  # Third part of the summary output of the Specificity function if not performing a subsequent randomization test
  specificity <- weighted_avg_same - weighted_avg_diff
  names(specificity) <- "specificity"
  
  # If the Specificity function is told to perform a randomization test, then only the specificity statistic is returned. The significance of this statistic is then calculated. 
  if (randomization == FALSE) {
    results <- list(avg_simils_df, weighted_avgs_df, specificity)
  } else {
    results <- specificity
  }

  return(results)
}
```
<br>

By default, the Specificity function will output a list containing three elements: i.) average similarity among all clone pairs, either evolved from the same parental genotype or between two parental genotypes (or antibiotic treatments); ii.) weighted averages; and iii.) the difference between these weighted averages, which is a measure of specificity of genomic evolution with respect to genetic background (or antibiotic treatment). Specifically, this calculation yields a positive or negative value that indicates the distinctness of clones evolved from the same parental genotype (or within the same treatment) relative to backgrounds from different parental genotypes (or between treatments). Similarity is higher in the former group and lower in the latter with increasing positive values. The converse is true if the value is negative (Sokal and Rohlf, 1994).
```{r Summary by genetic background}
# This map function requires that you pass the "antibiotics" vector to Specificity 
background_summary <- suppressMessages(map(antibiotics, ~ Specificity(specificity_bg, .x)))

names(background_summary) <- antibiotics
print(background_summary)
```
<br>

```{r Summary by antibiotic treatment}
antibiotic_summary <- suppressMessages(Specificity(specificity_ab))

print(antibiotic_summary)
```

<br>

If we consider an observed specificity statistic as one of many possible but equally likely different outcomes that could have arisen by chance, we can perform a randomization test to evaluate its significance (Sokal and Rohlf, 1994). To do this, we randomly rearrange the clones within each antibiotic treatment, but retain the number and identity of mutations in any clone. A similarity statistic is calculated for each trial and these outcomes are enumerated. Lastly, we quantify the proportion of trials that have a similarity statistic equal to, or greater than that observed (i.e., approximate *p*-value).
```{r Randomization test function, messages = FALSE}
# By default the RandomizationTest function will perform 10,000 trials. This value may be changed based upon your needs.
RandomizationTest <- function(dat, ab = NULL, trials = 10000) {
  trial_specificity <- c()
  
  orig_specificity <- Specificity(dat, ab)
  orig_specificity <- orig_specificity[[3]]$specificity
  
  for (i in 1:trials) {
    trial_specificity[i] <- Specificity(dat, ab, randomization = TRUE)
  }
  
  significance <- sum(trial_specificity >= orig_specificity) / trials # This equation gives the approximate p-value
    
  return(significance)
}
```
<br>

```{r Significance of specificity with respect to genetic background, messages = FALSE}
background_sig <- suppressMessages(
  map(antibiotics, ~ RandomizationTest(specificity_bg, .x)) 
  )

names(background_sig) <- antibiotics
background_sig
```
<br>

We performed an independent randomization test for each of the 4 antibiotics. These several independent tests bear upon the same null hypothesis that there will be an equivalent degree of similarity in mutational targets between replicates regardless of starting genetic background. Probabilities were thus combined from these independent significance tests using Fisher’s method with 8 degrees of freedom (i.e., df = 2k, where k is the number of comparisons) (Fisher 1934; Sokal and Rohlf 1994; Card et al. 2019)
```{r}
combined_prob <- sumlog(p = c(0.0435, 0.0132, 0.0907, 0.0846))
combined_prob
```


```{r Significance of specificity with respect to antibiotic treamtnet, messages = FALSE}
antibiotic_sig <- suppressMessages(
  RandomizationTest(specificity_ab)
)

print(antibiotic_sig)
```
<br>


***

# Figures

Summary of the mutations for each individual replicate by type of genetic change
```{r Mutations summary figure}
# Initial data wrangling
indv_mutations_summary_df <- mutations_curated_df %>%
  unite(col = "replicates", population:replicate, sep = "-") %>%
  group_by(replicates, antibiotic_resistance, mutation_type) %>% 
  summarize(count = n()) %>% 
  filter(!mutation_type == "Unknown")

# Function to create plots
plot_mutation_summary <- function(ab) {
  panel_plot <- indv_mutations_summary_df %>% 
    filter(antibiotic_resistance == ab) %>% 
    
    ggplot(aes(x = fct_rev(replicates), y = count, fill = mutation_type)) +
      geom_bar(stat = "identity") +
      coord_flip() +
      scale_fill_manual(breaks = c("Amplification > 50 bp", "Deletion > 50 bp", "Indel < 50 bp", "Intergenic", "IS insertion", "Nonsense",
                                   "Nonsynonymous", "Other", "Synonymous"),
                        values = c("#a50026", "#d73027", "#f46d43", "#fdae61", "#fee090", "#ffffbf", "#e0f3f8", "#abd9e9", "#4575b4"),
                        name = "Mutations") +
      theme_cowplot() +
      theme(plot.title = element_text(face = "plain", size = 12),
            axis.title.y = element_blank(),
            strip.background = element_blank(),
            strip.text = element_text(hjust = 0),
            legend.position = "none",
            plot.margin = margin(0, 0, -20, 0))
}

# Individual plots
p0 <- indv_mutations_summary_df %>%
  ggplot(aes(x = fct_rev(replicates), y = count, fill = mutation_type)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    facet_wrap(~fct_relevel(antibiotic_resistance, c("AMP", "CRO", "CIP", "TET")),
                labeller = as_labeller(antibiotic_labels), ncol = 1,
                scales = "free_y") + # To omit the 3 cross-contaminants
    ylab("Number of mutations") +
    scale_fill_manual(breaks = c("Amplification > 50 bp", "Deletion > 50 bp", "Indel < 50 bp", "Intergenic", "IS insertion", "Nonsense",
                                   "Nonsynonymous", "Other", "Synonymous"),
                        values = c("#a50026", "#d73027", "#f46d43", "#fdae61", "#fee090", "#ffffbf", "#e0f3f8", "#abd9e9", "#4575b4"),
                        name = "Mutations") +
    theme_cowplot() +
    theme(legend.position = "bottom",
          legend.justification = "center",
          axis.title.y = element_blank(),
          strip.background = element_blank(),
          strip.text = element_text(hjust = 0))

p1 <- plot_mutation_summary("AMP") + 
  labs(title = "Ampicillin") + 
  ylab(NULL) +
  theme(axis.line.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

p2 <- plot_mutation_summary("CRO") + 
  labs(title = "Ceftriaxone") + 
  ylab(NULL) +
  theme(axis.line.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

p3 <- plot_mutation_summary("CIP") + 
  labs(title = "Ciprofloxacin") + 
  ylab(NULL) + 
  scale_y_discrete(expand = expansion(add = 1.15), limits = c(0, 3)) +
  theme(axis.line.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

p4 <- plot_mutation_summary("TET") + 
  labs(title = "Tetracycline") + 
  ylab("Number of mutations") + 
  scale_x_discrete(limits = rev(unique(indv_mutations_summary_df$replicates)))

pcol <- plot_grid(p1, p2, p3, p4,
  align = 'vh',
  hjust = -1,
  ncol = 1
)

# extract a legend that is laid out horizontally
legend_b <- get_legend(
  p0 + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

# add the legend underneath the row we made earlier. Give it 10%
# of the height of one plot (via rel_heights).
indv_mutations_summary <- plot_grid(pcol, legend_b, ncol = 1, rel_heights = c(1, 0.1))
indv_mutations_summary

# ggsave("indv_mutations_summary.tif", indv_mutations_summary, path = "Figures", device = "tiff", width = 10, height = 14, units = "in", dpi = 300, compression = "lzw")
```
<br>


Parallelism at the functional level. 
```{r Targeted cellular processes}
cellular_processes <- poisson_analysis_final %>%
  ggplot(aes(x = process, y = observed_mutations)) +
    geom_col(fill = desaturate(lighten("#4575b4", .7), .5)) +
    geom_col(aes(y = expected_mutations), color = "black", alpha = 0.0, size = 1) +
    coord_flip() +
    ylab("Independent mutation events") +
    scale_x_discrete(limits = c("Replication", "Other", "Transcription or translation", "Transport", "Biosynthesis or metabolism", "Gene regulation")) +
    scale_y_continuous(limits = c(0, 25), breaks = c(0, 5, 10, 15, 20, 25)) +
    theme_cowplot() +
    theme(axis.title.y = element_blank())


cellular_processes

# ggsave("cellular_processes.tif", cellular_processes, path = "Figures", device = "tiff", width = 10, height = 5, units = "in", dpi = 300, compression = "lzw")
```
<br>

This figure shows all *qualifying* mutations in the sequenced genomes
```{r Mutation heatmap}
# Generates needed vectors
gene_order_vec <- mutations_curated_df %>% # Sorts genes by number of mutational hits and creates a vector
  filter(qualifying_mutation == 1) %>% 
  group_by(gene) %>% 
  summarize(hits = n()) %>% 
  arrange(desc(hits)) %>% 
  pull(gene)

replicate_names_vec <- mutations_curated_df %>%
  unite(col = "replicates", population:replicate, sep = "-") %>% 
  distinct(replicates) %>% 
  arrange(replicates) %>% 
  pull(replicates)

# Creates heatmap figure
heatmap <- specificity_bg %>%
  rename_at(vars(Anc_1:AraP5_3), ~ replicate_names_vec) %>% 
  pivot_longer(cols = 3:18, names_to = "replicate", values_to = "presence") %>% 
  mutate(presence = ifelse(presence == 0, "no", "yes")) %>% # Converts the numeric data to catagorical variables so I can use a discrete color scale in ggplot2
  filter(!(antibiotic == "AMP" & replicate == "Ara+4-2")) %>% # To omit the 3 cross-contaminants and 3 clones without "qualifying" mutations (see above)
  filter(!(antibiotic == "CRO" & replicate == "Ara+5-1")) %>% 
  filter(!(antibiotic == "CIP" & replicate %in% c("Ara+4-1", "Ara+4-3"))) %>% 
  
  ggplot(aes(x = fct_rev(replicate), y = gene, fill = presence)) +
    geom_tile(color = "white", size = 0.5) +
    coord_flip() +
    facet_wrap(~fct_relevel(antibiotic, c("AMP", "CRO", "CIP", "TET")),
              labeller = as_labeller(antibiotic_labels), ncol = 1,
              scales = "free_y") + 
    #scale_x_discrete(limits = rev(replicate_names_vec)) +
    scale_y_discrete(limits = gene_order_vec) +
    scale_fill_manual(
      breaks = c("no", "yes"),
      values = c(
      "no" = desaturate(lighten("#4575b4", .7), .5),
      "yes" = desaturate(darken("#4575b4", .4), .5)) 
    ) +
    theme_cowplot() +
    theme(axis.title = element_blank(),
          axis.text.x = element_text(face = "italic", angle = 90, hjust = 1, vjust = 0.5, color = "black"),
          axis.text.y = element_text(color = "black"),
          axis.ticks = element_blank(),
          axis.line = element_blank(),
          strip.background = element_blank(),
          strip.text = element_text(color = "black", hjust = 0),
          legend.position = "none")

heatmap

# ggsave("mutation_heatmap.tif", heatmap, path = "Figures", device = "tiff", width = 10, height = 14, units = "in", dpi = 300, compression = "lzw")
```



















