---
title: "LTEE ABR mutant sequencing notebook"
author: Kyle Card
Date: 2/1/2020
output: html_notebook
---

## Prerequisites

Clear current environment
```{r Clear current environment, messages = FALSE}
rm(list = ls())
```
<br>

Load packages for use
```{r Packages, messages = FALSE}
library(tidyverse)
library(reshape2)
library(proxy)
library(cowplot)
library(ggpubr)
library(cowplot)
```
<br>

***

```{r Factor levels}
strains <- c("KJC64", "KJC72", "KJC80", "KJC112", "KJC120", "KJC128", "KJC152", "KJC160", "KJC184", "KJC216", "KJC224", "KJC232", "KJC65", "KJC73", "KJC81", "KJC113", "KJC121", "KJC129", "KJC153", "KJC161", "KJC185", "KJC217", "KJC225", "KJC233", "KJC66", "KJC74", "KJC82", "KJC114", "KJC122", "KJC130", "KJC154", "KJC162", "KJC186", "KJC218", "KJC226", "KJC234", "KJC67", "KJC75", "KJC83", "KJC115", "KJC123", "KJC131", "KJC155", "KJC163", "KJC187", "KJC219", "KJC227", "KJC235")
antibiotics <- c("AMP", "CRO", "CIP", "TET")
antibiotic.labels <- c(
  `AMP` = "Ampicillin",
  `CRO` = "Ceftriaxone",
  `CIP` = "Ciprofloxacin",
  `TET` = "Tetracycline"
)
level.order <- as.factor(c("ompF", "ompR", "envZ", "alaT", "[ECB_00212]\u2013[phoE]", "gyrA", "hns", "rpoB", "agaA", "asnS", "[aspC]\u2013ompF", "aspC (ompF)", "baeR", "dnaG (rpoD)", "ECB_01487", "ftsI", "gumD", "hisQ", "hns (tdk)", "[lacY]\u2013yaiO", "lpcA", "marR", "ompR (greB)", "rpoD", "rrmA (cspC)", "slyA", "tuf", "waaC", "ybeY", "[ydaO]\u2013ynaE", "[yibD]\u2013[waaT]"))
```
<br>

***

# Data import and wrangling

```{r Data wrangling, messages = FALSE}
# Reads in the reference data obtained from Tenaillon et al. (2016)
references.df <- read.csv("Reference mutations.csv")
references.df$position <- as.character(references.df$position)

# Reads in the sample data
samples.df <- read.csv("Sample mutations.csv")
samples.df$position <- as.character(samples.df$position)

# Called mutations across all replicates and treatments, including their associated mutation type, annotation, target locus, cellular process effected, and description
mutations.df <- read.csv("Mutations.csv")

# Binary data showing whether a mutation is present (1) or absent (0) in a given locus across all replicates and treatments
mut.presence.df <- read.csv("Mutation presence - background specificity.csv")

mp.antibiotic.df <- read.csv("Mutation presence - ab specificity.csv")

# Mutation spectrum counts
spectrum.df <- mutations.df %>% 
  group_by(population, antibiotic.resistance, mutation.type) %>% 
  summarize(count = n())

# Mutation spectrum counts where the derived strains (Ara-5, Ara-6, Ara+4, and Ara+5) are combined into one group labeled "derived"
combined.spectrum.df <- spectrum.df
combined.spectrum.df$population <- fct_other(combined.spectrum.df$population, keep = "Ancestor", other_level = "Derived")

combined.spectrum.df <- combined.spectrum.df %>% 
  group_by(population, antibiotic.resistance, mutation.type) %>% 
  summarize(sum.counts = sum(count))
```
<br>

***

# Curation of *breseq* results

Each sequenced resistant isolate contains data on background and resistance mutations that evolved during the LTEE and our previous selection experiments (Card et al., 2019), respectively. We therefore discard the background mutations by comparing each isolate (sample) to its corresponding reference genome (Tenaillon et al., 2016). For example, isolate *KJC64* is compared to the *Ara+4 background at 50,000 generations*. If both the sample and reference have a mutation at the same genome position, then it is assumed that this is a background mutation and is discarded from further analyses. We further curated this list of mutational differences (*mutation.differences* object) by removing called sample mutations that are in deleted regions of the reference genome (possibly caused by differences in coverage between our two studies).    
```{r Filter mutations}
CompareMutations <- function(ref.data, sample.data) {
  
  compare.vec <- c()
  
  for(pop in unique(references.df$population)) {
    ref.subset <- ref.data %>% 
      filter(population == pop) %>% 
      select(position)
    
    sample.pop.subset <- sample.data %>% 
      filter(population == pop)
    
    for(strain in unique(sample.pop.subset$designation)) {
      sample.subset <- sample.pop.subset %>% 
        filter(designation == strain) %>% 
        select(position)
      
      compare.vec <- c(compare.vec, setdiff(sample.subset, ref.subset))
    }
  }
  return(compare.vec)
}

mutation.differences <- CompareMutations(references.df, samples.df)
names(mutation.differences) <- strains
```
<br>

***

# Specificity of genomic evolution with respect to background

We quantify the extent of parallelism in genome evolution between resistant mutants evolved from the same parental genotype and across different parental genotypes. We estimate Dice's coefficient of similarity, *S*, for each pair of evolved backgrounds, where $$S = \frac{2|X \cap Y|}{|X|+|Y|}$$ $|X|$ and $|Y|$ are the sets of genes with mutations in two evolved backgrounds, and $|X \cap Y|$ is the set of genes with mutations in both backgrounds. *S* therefore ranges from 0, when the pair of backgrounds share no mutations in common, to 1, when both have mutations in exactly the same set of genes (Deatherage et al., 2017).
```{r Genetic background specificity, messages = FALSE}
BGSpecificity <- function(dat, ab, randomization = FALSE) {
  mutations <- dat %>% 
    filter(antibiotic == ab) %>% 
    select(-(antibiotic:gene)) # Removes the first two columns, which are not needed in the function that estimates pairwise similarities
  
  if (randomization == TRUE) {
    colnames(mutations) <- sample(colnames(mutations))
  } else {
    mutations
  }
  
  # Computes similarity coefficient for all possible background pairs. Populates a matrix with these values
  pwise.simil.mat <- as.matrix(simil(mutations, method = "Dice", by_rows = FALSE, upper = TRUE, diag = TRUE))
  
  # All values above (and including) the matrix diagonal are converted to NA to ease downstream wrangling
  pwise.simil.mat[upper.tri(pwise.simil.mat)] <- NA
  
  # Converted matrix to a data frame for analysis
  pwise.simil.df <- melt(pwise.simil.mat, varnames = c("background.1", "background.2"))
  
  # NA values above (and including) the matrix diagonal are incorporated into the newly formed dataframe. This piece of code drops the rows contanining NA, effectively retaining only those values *below* the matrix diagonal
  df.filtered <- pwise.simil.df %>% 
    drop_na()
  
  # These lines of code remove the "_X" suffixes from the genotype labels to ease downstream wrangling
  df.filtered$background.1 <- str_replace_all(df.filtered$background.1, c("_1|_2|_3|_4"), "") 
  df.filtered$background.2 <- str_replace_all(df.filtered$background.2, c("_1|_2|_3|_4"), "")
  
  # First part of the summary output of the BGSpecificity function if not performing subsequent randomization tests
  avg.simils <- df.filtered %>% 
    group_by(background.1, background.2) %>% 
    summarize(avg = mean(value))
  
  avg.simils.df <- melt(avg.simils)
  avg.simils.df <- avg.simils.df %>% 
    select(-variable)
  
  # This next set of code pools backgrounds into two groups: i.) those that evolved from the same parental genotype (S_s); and ii.) those that evolved from different parental genotypes (S_d).
  Anc.simil <- df.filtered %>% 
    filter(background.1 == "Anc" & background.2 == "Anc")
  
  AraM5.simil <- df.filtered %>% 
    filter(background.1 == "AraM5" & background.2 == "AraM5")
  
  AraM6.simil <- df.filtered %>% 
    filter(background.1 == "AraM6" & background.2 == "AraM6")
  
  AraP4.simil <- df.filtered %>% 
    filter(background.1 == "AraP4" & background.2 == "AraP4")
  
  AraP5.simil <- df.filtered %>% 
    filter(background.1 == "AraP5" & background.2 == "AraP5")
    
  simil.same.df <- bind_rows(Anc.simil, AraM5.simil, AraM6.simil, AraP4.simil, AraP5.simil)
  simil.diff.df <- anti_join(df.filtered, simil.same.df)
  
  pooled.avg.same <- simil.same.df %>% 
    summarize(avg = mean(value)) %>% 
    round(digits = 5)
  
  pooled.avg.diff <- simil.diff.df %>% 
    summarize(avg = mean(value)) %>% 
    round(digits = 5)
  
  # Second part of the summary output of the BGSpecificity function if not performing subsequent randomization tests
  group.col <- data.frame(group = c("Same", "Different"))
  bound.pooled.avgs <- bind_rows(pooled.avg.same, pooled.avg.diff)
  pooled.avgs.df <- bind_cols(group.col, bound.pooled.avgs)
  
  # Third part of the summary output of the BGSpecificity function if not performing a subsequent randomization test.
  specificity <- pooled.avg.same - pooled.avg.diff
  names(specificity) <- "specificity"
  
  # By default, BGSpecificity will output a list containing three elements: i.) average similarity (Ss or Sd) for each background pair (15 comparisons total); ii.) average Ss and Sd by pooling backgrounds into these two groups; and iii.) the difference between these pooled averages, which is a measure of specificity of genomic evolution with respect to genetic background. Specifically, this calculation yields a positive or negative value that indicates the distinctness of backgrounds derived from the same parental genotype relative to backgrounds from different parental genotypes. Similarity is higher in the former group and lower in the latter with increasing positive values. The converse is true if the value is negative (Sokal and Rohlf, 1994).
  
  # If the BGSpecificity function is told to perform a randomization test, then only the specificity statistic is returned. The significance of this statistic is then calculated. 
  if (randomization == FALSE) {
    results <- list(avg.simils.df, pooled.avgs.df, specificity)
  } else {
    results <- specificity
  }

  return(results)
}

bg.summary <- suppressMessages(
  map(antibiotics, function(x){BGSpecificity(mut.presence.df, x)})
  )

names(bg.summary) <- antibiotics
print(bg.summary)
```
<br>

If we consider an observed specificity statistic as one of many possible but equally likely different outcomes that could have arisen by chance, we can perform a randomization test to evaluate its significance. To do this, we randomly rearrange the genotype labels within each antibiotic treatment, but retain the number and identity of mutations in any clone. A similarity statistic is calculated for each trial and these outcomes are enumerated. Lastly, we quantify the proportion of trials that have a similarity statistic equal to, or greater than that observed (i.e., *p*-value). 
```{r Significance of genetic background specificity, messages = FALSE}
# By default the BGRandomizationTest function will perform 10,000 trials. This value may be changed based upon your needs. 
BGRandomizationTest <- function(dat, ab, trials = 10000) {
  trial.specificity <- c()
  
  orig.specificity <- BGSpecificity(dat, ab)
  orig.specificity <- orig.specificity[[3]]$specificity
  
  for (i in 1:trials) {
    trial.specificity[i] <- BGSpecificity(dat, ab, randomization = TRUE)
  }
  
  significance <- sum(trial.specificity >= orig.specificity) / trials # This equation gives the approximate p-value
    
  return(significance)
}

bg.specificity.sig <- suppressMessages(
  map(antibiotics, function(x){BGRandomizationTest(mut.presence.df, x)})
  )

names(bg.specificity.sig) <- antibiotics
print(bg.specificity.sig)
```
<br>

***

# Specificity of genomic evolution with respect to antibiotic treatment

We next compare the similarity of mutations between replicates within the same antibiotic treatment and across treatments to evaluate the specificity of genomic evolution with respect to drug selection pressure.
```{r Antibiotic treatment specificity, messages = FALSE}
ABSpecificity <- function(dat, randomization = FALSE) {
  if (randomization == TRUE) {
    colnames(dat) <- sample(colnames(dat))
  } else {
    dat
  }
  
  # Computes similarity coefficient for all possible background pairs. Populates a matrix with these values
  pwise.simil.mat <- as.matrix(simil(dat, method = "Dice", by_rows = FALSE, upper = TRUE, diag = TRUE))
  
  # All values above (and including) the matrix diagonal are converted to NA to ease downstream wrangling
  pwise.simil.mat[upper.tri(pwise.simil.mat)] <- NA
  
  # Converted matrix to a data frame for analysis
  pwise.simil.df <- melt(pwise.simil.mat, varnames = c("background.1", "background.2"))
  
  # NA values above (and including) the matrix diagonal are incorporated into the newly formed dataframe. This piece of code drops the rows contanining NA, effectively retaining only those values *below* the matrix diagonal
  df.filtered <- pwise.simil.df %>% 
    drop_na()
  
  # These lines of code remove the "X_Y_" prefixes from the genotype labels to ease downstream wrangling
  df.filtered$background.1 <- str_replace_all(df.filtered$background.1, c("_1_|_2_|_3_|_4_"), "_") 
  df.filtered$background.2 <- str_replace_all(df.filtered$background.2, c("_1_|_2_|_3_|_4_"), "_")
  
  df.filtered$background.1 <- str_replace_all(df.filtered$background.1, c("Anc_|AraM5_|AraM6_|AraP4_|AraP5_"), "") 
  df.filtered$background.2 <- str_replace_all(df.filtered$background.2, c("Anc_|AraM5_|AraM6_|AraP4_|AraP5_"), "")
  
  # First part of the summary output of the ABSpecificity function if not performing subsequent randomization tests
  avg.simils <- df.filtered %>% 
    group_by(background.1, background.2) %>% 
    summarize(avg = mean(value))
  
  avg.simils.df <- melt(avg.simils)
  avg.simils.df <- avg.simils.df %>% 
    select(-variable)
  
  # This next set of code pools backgrounds into two groups: i.) those from the same antibiotic (S_s); and ii.) those from different antibiotics (S_d).
  AMP.simil <- df.filtered %>% 
    filter(background.1 == "AMP" & background.2 == "AMP")
  
  CRO.simil <- df.filtered %>% 
    filter(background.1 == "CRO" & background.2 == "CRO")
  
  CIP.simil <- df.filtered %>% 
    filter(background.1 == "CIP" & background.2 == "CIP")
  
  TET.simil <- df.filtered %>% 
    filter(background.1 == "TET" & background.2 == "TET")
    
  simil.same.df <- bind_rows(AMP.simil, CRO.simil, CIP.simil, TET.simil)
  simil.diff.df <- anti_join(df.filtered, simil.same.df)
  
  pooled.avg.same <- simil.same.df %>% 
    summarize(avg = mean(value)) %>% 
    round(digits = 5)
  
  pooled.avg.diff <- simil.diff.df %>% 
    summarize(avg = mean(value)) %>% 
    round(digits = 5)
  
  # Second part of the summary output of the ABSpecificity function if not performing subsequent randomization tests
  group.col <- data.frame(group = c("Within", "Between"))
  bound.pooled.avgs <- bind_rows(pooled.avg.same, pooled.avg.diff)
  pooled.avgs.df <- bind_cols(group.col, bound.pooled.avgs)
  
  # Third part of the summary output of the BGSpecificity function if not performing a subsequent randomization test.
  specificity <- pooled.avg.same - pooled.avg.diff
  names(specificity) <- "specificity"
  
  # By default, ABSpecificity will output a list containing three elements: i.) average similarity (Ss or Sd) for each background pair (10 comparisons total); ii.) average Ss and Sd by pooling backgrounds into these two groups; and iii.) the difference between these pooled averages, which is a measure of specificity of genomic evolution with respect to antibiotic treatment. Specifically, this calculation yields a positive or negative value that indicates the distinctness of backgrounds evolved in the same treatment relative to backgrounds from different treatments. Similarity is higher in the former group and lower in the latter with increasing positive values. The converse is true if the value is negative (Sokal and Rohlf, 1994).
  
  # If the ABSpecificity function is told to perform a randomization test, then only the specificity statistic is returned. The significance of this statistic is then calculated. 
  if (randomization == FALSE) {
    results <- list(avg.simils.df, pooled.avgs.df, specificity)
  } else {
    results <- specificity
  }

  return(results)
}

ab.summary <- suppressMessages(ABSpecificity(mp.antibiotic.df))
print(ab.summary)

```
<br>

```{r Significance of treatment specificity, messages = FALSE}
# By default the ABRandomizationTest function will perform 10,000 trials. This value may be changed based upon your needs. 
ABRandomizationTest <- function(dat, trials = 10000) {
  trial.specificity <- c()
  
  orig.specificity <- ABSpecificity(dat)
  orig.specificity <- orig.specificity[[3]]$specificity
  
  for (i in 1:trials) {
    trial.specificity[i] <- ABSpecificity(dat, randomization = TRUE)
  }
  
  significance <- sum(trial.specificity >= orig.specificity) / trials # This equation gives the approximate p-value
    
  return(significance)
}

ab.specificity.sig <- suppressMessages(ABRandomizationTest(mp.antibiotic.df))
```


***
# Figures

```{r Figure 1}
GenerateFig1.abc <- function(dat, ab, panel.title) {
  ab.frame <- dat %>% 
    filter(antibiotic.resistance == ab)
  
  panel <- ggplot(ab.frame, aes(x = population, y = count, fill = type)) +
    geom_bar(position = "fill", stat = "identity") +
    coord_flip() +
    ggtitle(panel.title) +
    labs(y = "") +
    scale_fill_brewer(type = "div") +
    scale_x_discrete(limit = c("Ara+5", "Ara+4", "Ara\u20136", "Ara\u20135", "Ancestor")) +
    guides(fill = guide_legend(reverse = TRUE)) +
    theme(axis.title.y = element_blank(),
          axis.text.y = element_text(color = "black", size = 16),
          axis.text.x = element_text(color = "white", size = 16),
          axis.ticks = element_blank(),
          axis.line.x = element_blank(),
          plot.margin = unit(c(1, 1, 0, 1), "cm"),
          plot.title = element_text(color = "black", size = 16),
          legend.title = element_blank(),
          legend.text = element_text(color = "black", size = 16),
          legend.box.margin = margin(10, 10, 10, 10))
  
  return(panel)
  
}

fig1.abc <- map2(antibiotics, c("Ampicillin", "Ceftriaxone", "Ciprofloxacin"), function(x, y){GenerateFig1.abc(spectrum.df, x, y)})


fig1.d <- ggplot(spectrum.df %>% filter(antibiotic.resistance == "TET"),
                 aes(x = population, y = count, fill = type)) +
  geom_bar(position = "fill", stat = "identity") +
  coord_flip() +
  ggtitle("Tetracycline") +
  labs(y = "Frequency of mutations") +
  scale_fill_brewer(type = "div") +
  scale_x_discrete(limit = c("Ara+5", "Ara+4", "Ara\u20136", "Ara\u20135", "Ancestor")) +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme(axis.title.y = element_blank(),
        axis.title.x = element_text(color = "black", size = 16),
        axis.text = element_text(color = "black", size = 16),
        axis.ticks = element_blank(),
        plot.margin = unit(c(1, 1, 0.5, 1), "cm"),
        plot.title = element_text(color = "black", size = 16),
        legend.title = element_blank(),
        legend.text = element_text(color = "black", size = 16),
        legend.box.margin = margin(10, 10, 10, 10))


fig1 <- ggarrange(fig1.abc[[1]], fig1.abc[[2]], fig1.abc[[3]], fig1.d, nrow = 4, common.legend = TRUE, legend = "right")
print(fig1)

# ggsave("fig1.tif", fig1, path = "Figures", device = "tiff", width = 12, height = 10, units = "in", dpi = 300, compression = "lzw")
```
<br>

```{r Figure 2}
GenerateFig2.abc <- function(dat, ab, panel.title) {
  ab.frame <- dat %>% 
    filter(antibiotic.resistance == ab)
  
  panel <- ggplot(ab.frame, aes(x = population, y = count, fill = type)) +
    geom_bar(position = "fill", stat = "identity") +
    coord_flip() +
    ggtitle(panel.title) +
    labs(y = "") +
    scale_fill_brewer(type = "div") +
    scale_x_discrete(limit = c("Derived", "Ancestor")) +
    guides(fill = guide_legend(reverse = TRUE)) +
    theme_cowplot() +
    theme(axis.title.y = element_blank(),
          axis.text.y = element_text(color = "black", size = 16),
          axis.text.x = element_text(color = "white", size = 16),
          axis.ticks = element_blank(),
          plot.margin = unit(c(1, 1, 0, 1), "cm"),
          plot.title = element_text(color = "black", size = 16),
          legend.title = element_blank(),
          legend.text = element_text(color = "black", size = 16),
          legend.box.margin = margin(10, 10, 10, 10))
  
  return(panel)
  
}

fig2.abc <- map2(antibiotics, c("Ampicillin", "Ceftriaxone", "Ciprofloxacin"), function(x, y){GenerateFig2.abc(combined.spectrum.df, x, y)})


fig2.d <- ggplot(combined.spectrum.df %>% filter(antibiotic.resistance == "TET"),
                 aes(x = population, y = count, fill = type)) +
  geom_bar(position = "fill", stat = "identity") +
  coord_flip() +
  ggtitle("Tetracycline") +
  labs(y = "Frequency of mutations") +
  scale_fill_brewer(type = "div") +
  scale_x_discrete(limit = c("Derived", "Ancestor")) +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme_cowplot() +
  theme(axis.title.y = element_blank(),
        axis.title.x = element_text(color = "black", size = 16),
        axis.text = element_text(color = "black", size = 16),        
        axis.ticks = element_blank(),
        plot.margin = unit(c(1, 1, 0.5, 1), "cm"),
        plot.title = element_text(color = "black", size = 16),
        legend.title = element_blank(),
        legend.text = element_text(color = "black", size = 16),
        legend.box.margin = margin(10, 10, 10, 10))


fig2 <- ggarrange(fig2.abc[[1]], fig2.abc[[2]], fig2.abc[[3]], fig2.d, nrow = 4, common.legend = TRUE, legend = "right")
print(fig2)

# ggsave("fig2.tif", fig2, path = "Figures", device = "tiff", width = 12, height = 10, units = "in", dpi = 300, compression = "lzw")
```
<br>


```{r Figure 3}
GenerateFig3.bcd <- function(dat, ab, panel.title) {
  ab.frame <- dat %>% 
    filter(antibiotic.resistance == ab)
  
  panel <- ggplot(ab.frame, aes(x = population, y = gene, fill = mutation.type)) +
    geom_tile() +
    ggtitle(panel.title) +
    scale_fill_brewer(type = "div") +
    scale_y_discrete(limits = rev(level.order)) +
    scale_x_discrete(limit = replicates) +
    guides(fill = guide_legend(mutation.type, reverse = TRUE)) +
    theme(axis.title = element_blank(),
          axis.text.y = element_blank(),
          axis.text.x = element_text(color = "black", size = 16, angle = 90, hjust = 1, vjust = 0.5),
          axis.ticks = element_blank(),
          panel.background = element_rect(color = "#DCDCDC", fill = "#DCDCDC"),
          plot.title = element_text(color = "black", size = 16),
          legend.title = element_blank(),
          legend.text = element_text(color = "black", size = 16),
          legend.box.margin = margin(10, 10, 10, 10))
  
  return(panel)
  
}


fig3.bcd <- map2(c("CRO", "CIP", "TET"), c("Ceftriaxone", "Ciprofloxacin", "Tetracycline"), function(x, y){GenerateFig3.bcd(mutations.df, x, y)})


fig3.a <- ggplot(mutations.df %>% filter(antibiotic.resistance == "AMP"),
                 aes(x = population, y = gene, fill = mutation.type)) +
  geom_tile() +
  ggtitle("Ampicillin") +
  scale_fill_brewer(type = "div") +
  scale_y_discrete(limits = rev(level.order)) +
  scale_x_discrete(limit = replicates) +
  guides(fill = guide_legend(mutation.type, reverse = TRUE)) +
  theme(axis.title = element_blank(),
        axis.text.y = element_text(face = "italic", colour = "black", size = 16),
        axis.text.x = element_text(color = "black", size = 16, angle = 90, hjust = 1, vjust = 0.5),
        axis.ticks = element_blank(),
        panel.background = element_rect(color = "#DCDCDC", fill = "#DCDCDC"),
        plot.title = element_text(color = "black", size = 16),
        legend.title = element_blank(),
        legend.text = element_text(color = "black", size = 16),
        legend.box.margin = margin(10, 10, 10, 10))

fig3 <- ggarrange(fig3.a, fig3.bcd[[1]], fig3.bcd[[2]], fig3.bcd[[3]], ncol = 4, common.legend = TRUE, legend = "bottom", widths = c(1.445, 1, 1, 1))
print(fig3)

# ggsave("fig3.tif", fig3, path = "Figures", device = "tiff", width = 20, height = 10, units = "in", dpi = 300, compression = "lzw")
```



```{r}
ab.labels <- c(
  `AMP` = "Ampicillin",
  `CRO` = "Ceftriaxone",
  `CIP` = "Ciprofloxacin",
  `TET` = "Tetracycline"
)

ggplot(mutations.df, aes(x = population, y = gene, fill = mutation.type)) +
  geom_tile() +
  facet_wrap(~antibiotic.resistance, labeller = as_labeller(ab.labels)) +
  theme(axis.text.x = element_text(size = 16, angle = 90, hjust = 1, vjust = 0.5))
```
















